.ci/ReleaseChangelog.java

```diff
@@ -0,0 +1,295 @@
+/*
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import java.io.IOException;
+import java.nio.charset.StandardCharsets;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.time.LocalDate;
+import java.time.format.DateTimeFormatter;
+import java.util.ArrayList;
+import java.util.Comparator;
+import java.util.List;
+import java.util.Locale;
+import java.util.OptionalInt;
+import java.util.function.Predicate;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+public class ReleaseChangelog {
+
+  public static void main(String[] args) throws IOException {
+    if (args.length != 3) {
+      System.out.println(
+          "Expected exactly three arguments: <ChangelogFile> <ReleaseNotesPath> <VersionToRelease>");
+      System.exit(-1);
+    }
+    Path nextChangelogFile = Paths.get(args[0]);
+    Path releaseNotesDir = Paths.get(args[1]);
+    Path releaseNotesFile = releaseNotesDir.resolve("index.md");
+    Path deprecationsFile = releaseNotesDir.resolve("deprecations.md");
+    Path breakingChangesFile = releaseNotesDir.resolve("breaking-changes.md");
+    VersionNumber version = VersionNumber.parse(args[2].trim());
+
+    Lines nextChangelogLines = new Lines(
+        Files.readAllLines(nextChangelogFile, StandardCharsets.UTF_8));
+    Lines fixes = nextChangelogLines.cutLinesBetween("<!--FIXES-START-->", "<!--FIXES-END-->");
+    Lines enhancements = nextChangelogLines.cutLinesBetween("<!--ENHANCEMENTS-START-->",
+        "<!--ENHANCEMENTS-END-->");
+    Lines deprecations = nextChangelogLines.cutLinesBetween("<!--DEPRECATIONS-START-->",
+        "<!--DEPRECATIONS-END-->");
+    Lines breakingChanges = nextChangelogLines.cutLinesBetween("<!--BREAKING-CHANGES-START-->",
+        "<!--BREAKING-CHANGES-END-->");
+
+    Lines dependenciesNotes = nextChangelogLines.cutLinesBetween("<!--DEPENDENCIES-NOTES-START-->",
+        "<!--DEPENDENCIES-NOTES-END-->");
```

Comment:
[for reviewer] this section does not exists in the next release changelog template, we have to append it manually at runtime. This is due to the fact that we can't easily invoke the gradle task from here, doing everything from gradle in the future would simplify this.